### PR TITLE
Create new endpoints that allow to add/update/edit project with dependency network

### PR DESCRIPTION
**Pull request by hoangnt2**
## Create new endpoints that allow to add/update/edit project with dependency network

**`I have split this PR to separate commits, that will make it easier for reviewers to review. (Will squash commits after review successfully if needed)`**

# This PR contains 6 new endpoints:
1. Get linked releases in dependency network of a project
2. Get linked releases of a release
3. Get linked releases from sub projects of a project
4. Compare project network (relationship in project network) with default network (relationship between release in release table)
5. Check cyclic links between releases
6. Duplicate project with network

# Pre-condition:
Turn on feature by adding this property into **/etc/sw360/sw360.properties**: **enable.flexible.project.release.relationship=true**

**Create a new project with dependency network**:
- Url: http://localhost:8080/resource/api/projects/network
- Method: POST
- Body: 
```
{
    "name": "Test_PR",
    "dependencyNetwork": [
        {
            "releaseId": "${release_id_1_in_sw360}", // Replace this by an existing release id in sw360
            "releaseRelationship": "CONTAINED",
            "mainlineState": "OPEN",
            "comment": ""
            "releaseLink": [
                {
                    "releaseId": "${release_id_2_in_sw360}", // Replace this by an existing release id in sw360 (difference from above)
                    "releaseRelationship": "OPTIONAL",
                    "mainlineState": "MAINLINE",
                    "comment": "",
                    "releaseLink": []
                }
            ]
        }
    ]
}
```

# How To Test?

### 1. Get linked releases in dependency network of a project**
- Url: http://localhost:8080/resource/api/projects/network/{project_id}/linkedReleases
- Method: GET
- Expected output:

```
[
    {
        "releaseId": "${release_id_1_in_sw360}",
        "releaseName": "${release_1_name}",
        "releaseVersion": "${release_1_version}",
        "componentId": "${release_1_component_id}",
        "releaseRelationship": "CONTAINED",
        "mainlineState": "OPEN",
        "comment": "",
        "releaseLink": [
            {
                "releaseId": "${release_id_2_in_sw360}",
                "releaseName": "${release_2_name}",
                "releaseVersion": "${release_2_version}",
                "componentId": "${release_2_component_id}",
                "releaseRelationship": "OPTIONAL",
                "mainlineState": "MAINLINE",
                "comment": "",
                "releaseLink": []
            }
        ]
    }
]
```

### 2. Get linked releases of a release
- Url: http://localhost:8080/resource/api/releases/{release_id}/releases
- Method: GET
- Expected output:
```
{
    "_embedded": {
        "sw360:releaseLinks": [
            {
                "id": "${linked_release_id}",
                "name": "${linked_release_name}",
                "version": "${linked_release_version}",
                "releaseRelationship": "CONTAINED",
                "clearingState": "NEW_CLEARING",
                "licenseIds": [],
                "accessible": true,
                "componentId": "${linked_release_component_id}"
            }
        ]
    },
    "_links": {
        "curies": [
            {
                "href": "http://localhost:8080/resource/docs/{rel}.html",
                "name": "sw360",
                "templated": true
            }
        ]
    }
}
```

### 3. Get linked releases from sub projects of a project
**Step 1: Create a parent project of create Test_PR (in pre-condition):**
- URL: http://localhost:8080/resource/api/projects
- Method: POST
- Body:
```
{
    "name": "Test_PR_Parent_Project",
    "linkedProjects": {
        "{created_test_project_id}": {
            "projectRelationship": "CONTAINED",
            "enableSvm": true
        }
    }
}
```

**Step 2: Get linked releases of sub projects (Test_PR) from parent project (Test_PR_Parent_Project):**
- URL: http://localhost:8080/resource/api/projects
- Method: GET
- Expected Output: Return directly linked release of Test_PR as following format
```
{
    "_embedded": {
        "sw360:releases": [
            {
                "id": "${release_id_1_in_sw360}",
                "cpeid": "",
                "name": "${release_name}",
                "version": "${release_version}",
                "releaseDate": "",
                "sourceCodeDownloadurl": "",
                "binaryDownloadurl": "",
                "mainlineState": "OPEN",
                "clearingState": "NEW_CLEARING",
                "externalToolProcesses": [],
                "_links": {
                    "sw360:component": {
                        "href": "http://localhost:8080/resource/api/components/${release_component_id}"
                    },
                    "self": {
                        "href": "http://localhost:8080/resource/api/releases/${release_id_1_in_sw360}"
                    }
                }
            }
        ]
    },
    "_links": {
        "curies": [
            {
                "href": "http://localhost:8080/resource/docs/{rel}.html",
                "name": "sw360",
                "templated": true
            }
        ]
    }
}
```

### 4.  Compare project network (relationship in project network) with default network (relationship between release in release table)
- URL: http://localhost:8080/resource/api/projects/network/compareDefaultNetwork
- Method: POST
- Body:
```
 [
        {
            "releaseId": "${release_id_1_in_sw360}", // Replace this by an existing release id in sw360
            "releaseRelationship": "CONTAINED",
            "mainlineState": "OPEN",
            "comment": ""
            "releaseLink": [
                {
                    "releaseId": "${release_id_2_in_sw360}", // Replace this by an existing release id in sw360 (difference from above)
                    "releaseRelationship": "OPTIONAL",
                    "mainlineState": "MAINLINE",
                    "comment": "",
                    "releaseLink": []
                }
            ]
        }
    ]
```

- Expected output:

```
 [
        {
            "releaseId": "${release_id_1_in_sw360}", // Replace this by an existing release id in sw360
            "releaseRelationship": "CONTAINED",
            "mainlineState": "OPEN",
            "comment": ""
            "isDiff": false,
            "releaseLink": [
                {
                    "releaseId": "${release_id_2_in_sw360}", // Replace this by an existing release id in sw360 (difference from above)
                    "releaseRelationship": "OPTIONAL",
                    "mainlineState": "MAINLINE",
                    "comment": "",
                    "releaseLink": [],
                    "isDiff": true,
                }
            ]
        }
    ]
```

### 5. Check cyclic links between releases
- URL: http://localhost:8080/resource/api/releases/{release_id_1}/checkCyclicLink
- Method: POST
- Body:
```
{
    "linkedToReleases": ["{release_id_1}"],//replace by release id in sw360
    "linkedReleases": ["{release_id_2}"] //replace by release id in sw360
}
```
- Example response format:
```
[
    {
        "message": "m (1) -> m (1)",
        "status": 409
    },
    {
        "message": "There are no cyclic link between 9649fd4f4c7b45849c40aad08f1a625a and 0a4d467b19884d79b53251646c6e7393",
        "status": 200
    }
]
```

### 6. Duplicate project with network:
- URL: http://localhost:8080/resource/api/projects/network/duplicate/{created_project_id_in_pre_condition}
- Method: POST:
- Body:
```
{
    "name": "Test_PR_Duplicate"
}
```
- Expected: Dependency network is copied successfully into duplicated project
```
{
    "name": "Test_PR_Duplicate",
    "createdOn": "2024-09-13",
    "businessUnit": "DEPARTMENT",
    "state": "ACTIVE",
    "clearingState": "OPEN",
    "securityResponsibles": [],
    "enableSvm": false,
    "considerReleasesFromExternalList": false,
    "enableVulnerabilitiesDisplay": false,
    "dependencyNetwork": [
        {
            "releaseId": "${release_id_1_in_sw360}",
            "releaseRelationship": "CONTAINED",
            "mainlineState": "OPEN",
            "comment": "",
            "createOn": "2024-09-13",
            "createBy": "admin@sw360.org",
            "releaseLink": [
                {
                    "releaseId": "${release_id_2_in_sw360}",
                    "releaseRelationship": "OPTIONAL",
                    "mainlineState": "MAINLINE",
                    "comment": "",
                    "createOn": "2024-09-13",
                    "createBy": "admin@sw360.org",
                    "releaseLink": []
                }
            ]
        }
    ],
    "projectType": "PRODUCT",
    "visibility": "EVERYONE",
    "_links": {
        "self": {
            "href": "http://localhost:8080/resource/api/projects/4d13f3f5ec1e49fca4986ebbf5e4f891"
        }
    },
    "_embedded": {
        "createdBy": {
            "email": "user@sw360.org",
            "wantsMailNotification": false,
            "deactivated": false,
            "_links": {
                "self": {
                    "href": "http://localhost:8080/resource/api/users/byid/sw360-user"
                }
            }
        }
    }
}
```

### CODE REVIEW

File: libraries/datahandler/src/main/thrift/components.thrift [hunk 1]
```diff
@@ -480,13 +481,16 @@ struct ClearingReport{
...
-     3: optional string mainlineState,
-     4: optional string comment,
-     5: optional string createOn,
-     6: optional string createBy,
-     7: optional list<ReleaseNode> releaseLink,
+    1: required string releaseId,
+    8: optional string releaseName, // This field is used to view data, is not stored in db
+    9: optional string releaseVersion, // This field is used to view data, is not stored in db
+    10: optional string componentId, // This field is used to view data, is not stored in db
+    2: optional string releaseRelationship,
```
Comment (GMishx) [inline libraries/datahandler/src/main/thrift/components.thrift:488]: Please rearrange the fields as per their declaration order (8, 9, 10 after 7) :-)

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java [hunk 2]
```diff
@@ -1282,6 +1286,34 @@ public ResponseEntity<?> loadSpdxLicensesInfo(
         return new ResponseEntity<>(responseBody, HttpStatus.OK);
     }
 
+    @GetMapping(value = RELEASES_URL + "/{id}/releases")
+    public ResponseEntity<CollectionModel<HalResource<ReleaseLink>>> getLinkedReleases(
+            @PathVariable("id") String id,
+            @RequestParam(value = "transitive", required = false, defaultValue = "false") boolean transitive
+    ) throws TException {
```
Comment (GMishx) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java:1293]: Please add relevant `@Operation` and `@Parameter` annotations for correctly document the endpoint in OpenAPI doc.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java [hunk 3]
```diff
@@ -1353,6 +1385,76 @@ public ResponseEntity loadAssessmentSummaryInfo(
         return new ResponseEntity<>(assessmentSummaryMap, HttpStatus.OK);
     }
 
+    @Operation(
+            summary = "Check cyclic hierarchy of a release with other releases.",
+            description = "Check cyclic hierarchy of a release with other releases."
+    )
```
Comment (GMishx) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java:1391]: Please add the `Releases` tag to group endpoint correctly.
```suggestion
    @Operation(
            summary = "Check cyclic hierarchy of a release with other releases.",
            description = "Check cyclic hierarchy of a release with other releases.",
            tags = {"Releases"}
    )
```

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java [hunk 4]
```diff
@@ -1353,6 +1385,76 @@ public ResponseEntity loadAssessmentSummaryInfo(
...
+    @RequestMapping(value = RELEASES_URL + "/{id}/checkCyclicLink", method = RequestMethod.POST)
+    public ResponseEntity<?> checkForCyclicReleaseLink(
+            @Parameter(description = "The ID of the checking release.")
+            @PathVariable("id") String releaseId,
+            @RequestBody Map<String, Set<String>> relationshipReleaseIds
```
Comment (GMishx) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java:1396]: Please add example for this in OpenAPI docs. A sample can be:

```java
@Parameter(description = "Releases to check",
        schema = @Schema(example = """
                {
                  "linkedReleases": ["3765276512", "987456"],
                  "linkedToReleases": ["12121212"]
                }""")
)
```

**Review by GMishx — CHANGES_REQUESTED** [review]
Minor documentation changes needed.

File: libraries/datahandler/src/main/thrift/components.thrift (hunk 1 above)
Comment (hoangnt2) [reply libraries/datahandler/src/main/thrift/components.thrift]: @GMishx thanks for your comment, I fixed it

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java (hunk 4 above)
Comment (hoangnt2) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java:1470]: @GMishx thanks for your comment, I fixed it

**Review by GMishx — APPROVED** [review]
Changes looks good.

**Pull request merged**